### PR TITLE
Add a doc about Floating Point math as background on numeric testing

### PR DESCRIPTION
**Pull request by zoddicus**
Issue #1846

### CODE REVIEW

**Comment by zoddicus** [comment]
Please review for content, i.e. am I correct in my description of fp matters, and for formatting, i.e. is there a better way to display this in markdown.

File: docs/fp_primer.md [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+
+## Accuracy
+
+As mentioned above floating point numbers are not able to represent all of the
+possible values over their bounds, but instead represent discrete values in that
+interval, and approximate the remainder.
+
+Additionally, floating point numbers are not evenly distributed over their in 
+interval, but instead are clustered closer together near zero, and further apart
+near the edges.
```
Comment (dneto0) [inline docs/fp_primer.md:193]: "edges" is ambiguous?  Like,  they're bunched closer to the bottom edge of the interval [1000, 1000000].

I think you mean over the whole finite range of floats, so:

> ... are note evenly distributed over the real number line, but instead are ..  .. and further apart as their magnitudes grow.

File: docs/fp_primer.md [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+   is always expected to be exactly representable. This doesn't happen for any
+   of the operations that return floating point values, but does occur for
+   logical operations that return boolean values.
+
+
+2. *Correctly Rounded*
+
+   For the case that the true value is exactly representable as a f32, this is
+   the equivalent of exactly from above. In the event that the true value is not
+   exact, then the acceptable answer for most numbers is the nearest f32 above
```
Comment (dneto0) [inline docs/fp_primer.md:229]: "the nearest" sounds like one answer

> .... for most numbers is either the nearest f32 above or the nearest f32 below the true value.

File: docs/fp_primer.md [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+   and below the true value. 
+
+   For values near the subnormal range, e.g. close to zero, this becomes more
+   complex, since an implementation may FTZ at any point. So if the exact 
+   solution is subnormal or either of the neighbours of the true value are
+   subnormal, zero becomes a possible result, thus the acceptance interval is
```
Comment (dneto0) [inline docs/fp_primer.md:235]: Ah, this is the voice of hard-won experience.  :-)


No request for change here. Just acknowledging.

File: docs/fp_primer.md [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+4. *Units of Least Precision (ULP)*
+
+   The solution to the issue of not scaling with precision of floating point is
+   to use units of least precision.
+
+   There are many more formal definitions of ULP, but at a high level it is the
+   difference from a specific value to the next nearest f32 value.
+
+   n * ULP or nULP means `[X - n * ULP @ X, X + n * ULP @ X]`.
+
```
Comment (dneto0) [inline docs/fp_primer.md:260]: The geek in me begs for the formal definition of ULP here, after statement of n*ULP.  This is the place for being nerdly-correct.


> ULP(X) is   min (b-a) over all pairs (a,b) of representable floating point numbers such that (a <= X <= b)

File: docs/fp_primer.md [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+2. *Correctly Rounded*
+   
+   If `X` is precisely defined as a f32
+
+   `f(x) => [X, X]`
+
+   otherwise,
+
+   `[a, b]` where `a` is `min(X - a)`, and `b` is `min(b - X)`
```
Comment (dneto0) [inline docs/fp_primer.md:313]: The min expressions are confusing.

Should it be 

> Where `a` is the largest representable number with a <= X, and `b` is the smallest representable number with X <= `b`

File: docs/fp_primer.md [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+sometimes involve mapping domain intervals to range intervals, i.e. 
+`f([a, b]) = [A, B]`.
+
+Given that floating point numbers have a finite number of precise values for any
+given interval, one could implement just running the accuracy computation for
+every point in the interval and then spanning together the resultant intervals.
+That would be very inefficient though and make your reviewer sad to read.
+
+For mapping intervals to intervals the key insight is that we only need to be
+concerned with the extremas of the operation in the interval, since the
```
Comment (dneto0) [inline docs/fp_primer.md:338]: "extrema" is already plural.

File: docs/fp_primer.md [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+4. **ULP Error**
+
+   `f(x) = X => [X - n*ULP(X), X + n*ULP(X)]`
+
+As defined, these definitions handle mapping from a point in the domain into an
+interval in the range.
+
+This is insufficient for implementing inherited accuracies, since inheritance
+sometimes involve mapping domain intervals to range intervals, i.e. 
+`f([a, b]) = [A, B]`.
```
Comment (dneto0) [inline docs/fp_primer.md:330]: > Here we use the convention for naturally extending a function on real numbers into a function on *intervals* of real numbers.

File: docs/fp_primer.md [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+In more precise terms:
+```
+  f(x) => X, x = [a, b] and X = [A, B]
+  
+  X = [min(f(x)), max(f(x))]
+  X = [min(f([a, b])), max(f([a, b]))]
+  X = [f(m), f(M)]
+```
+where m and M are in `[a, b]`, `m <= M`, and produce the min and max results
+respectively.
```
Comment (dneto0) [inline docs/fp_primer.md:350]: >  and produce the min and max results for `f` on the interval, respectively.

File: docs/fp_primer.md [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+
+So how do we find the minima and maxima for our operation in the domain?
+
+The general solution for this requires using calculus to calculate the
```
Comment (dneto0) [inline docs/fp_primer.md:354]: > Generally, the solution for this....

I want to exclude the `step` function, for example, which is not differentiable at the 'edge' value.  :-)

File: docs/fp_primer.md [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+derivative of `f`, `f'`, and then find the zeroes `f'` to find inflection 
+points of `f`.
+
+Thankfully we do not need a general solution for the CTS, since all the builtin
+operations are defined in the spec, so `f` is from a known set of options.
+
+These operations can be divided into two broad categories: monotonic, and 
+non-monotonic.
```
Comment (dneto0) [inline docs/fp_primer.md:362]: this is always with respect to an interval (which becomes important later, I think).

So

> ... monotonic, and non-monotonic, with respect to an interval.

File: docs/fp_primer.md [hunk 11]
```diff
@@ -0,0 +1,495 @@
...
+
+The monotonic operations are ones that preserve the order of inputs in their
+outputs (or reverse it). Their graph only ever decreases or increases, 
+never changing from one or the other, though it can have flat sections.
+
+The non-monotonic operations are ones whose graph would have both regions of
+increase and decrease.
+
+The monotonic operations, when mapping an interval to an interval, are simple to
+do, since the extrema are guaranteed to be the ends of the domain, `a` and `b`.
```
Comment (dneto0) [inline docs/fp_primer.md:372]: s/do/handle/

File: docs/fp_primer.md [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+or `[f(b), f(a)]`. 
+
+It is simpler to just use min & max to have an implementation that is agnostic
+to the details of `f`.
+```
+  A = f(a), B = f(b)
+  X = [min(A, B), max(A, B)]
+```
+
+The non-monotonic functions that we need to handle interval to interval mappings
```
Comment (dneto0) [inline docs/fp_primer.md:385]: > we need to handle for interval-to-interval mappings ...

File: docs/fp_primer.md [hunk 13]
```diff
@@ -0,0 +1,495 @@
...
+are more complex, but thankfully are a small number of the overall operations
```
Comment (dneto0) [inline docs/fp_primer.md:386]: Suggest cutting off the sentence:

> are more complex.  Thankfully, they are small subset of the overall operations ....

File: docs/fp_primer.md [hunk 14]
```diff
@@ -0,0 +1,495 @@
...
+that need to be handled, since they are only the operations that used in an
```
Comment (dneto0) [inline docs/fp_primer.md:387]: "that are used in"

File: docs/fp_primer.md [hunk 15]
```diff
@@ -0,0 +1,495 @@
...
+inherited accuracy and take in the output of another operation as part of that
+inherited accuracy.
+
+So in the CTS we just have bespoke implementations for each of them. 
+
+Part of the operation definition in the CTS is a function that takes in the
+domain interval and returns an adjusted domain interval with the minima and
```
Comment (dneto0) [inline docs/fp_primer.md:394]: Can we be more descriptive here:

> takes in the domain interval, and returns a subinterval such that the subject function is monotonic over that subinterval, and hence the function's minima and maxima are at the ends.

File: docs/fp_primer.md [hunk 16]
```diff
@@ -0,0 +1,495 @@
...
+```
+  x = π/2
+
+  sin(π/2) => [sin(π/2) - 2-11, sin(π/2) + 2-11]
+           => [0 - 2-11, 0 + 2-11]
+           => [-0.000488.., 0.000488...]
+  cos(π/2) => [cos(π/2) - 2-11, cos(π/2) + 2-11]
+           => [-0.500488, -0.499511...]
+
+  tan(π/2) => sin(π/2)/cos(π/2)
```
Comment (dneto0) [inline docs/fp_primer.md:439]: This is illustrative!

File: docs/fp_primer.md [hunk 17]
```diff
@@ -0,0 +1,495 @@
...
+           => [0.000488.../-0.499511..., 0.000488.../0.499511...]
+           => [-0.0009775171, 0.0009775171]
+```
+
+For clarity this has omitted a bunch of complexity around FTZ behaviours, and
+that these operations are only defined for specific domains, but the high-level
+concepts hold.
+
+For each of the inherited operations we could implement a manually written out
+closed form solution like, but that would be quite error-prone and not be
```
Comment (dneto0) [inline docs/fp_primer.md:452]: drop "like" here?

File: docs/fp_primer.md [hunk 18]
```diff
@@ -0,0 +1,495 @@
...
+re-using code between builtins.
+
+Instead, the CTS takes advantage of the fact that for testing implementations
+version of `tan(x)`we are going to need tests for `sin(x)`, `cos(x)` and `x/y`,
```
Comment (dneto0) [inline docs/fp_primer.md:456]: need a space before "we"

**Review by dneto0 — CHANGES_REQUESTED** [review]
This is a very nice walkthrough!

Feedback is mainly minor edits, often with suggested changes.

File: docs/fp_primer.md (hunk 3 above)
Comment (zoddicus) [reply docs/fp_primer.md:242]: :+1:

File: docs/fp_primer.md (hunk 4 above)
Comment (zoddicus) [reply docs/fp_primer.md:268]: Done

File: docs/fp_primer.md (hunk 8 above)
Comment (zoddicus) [reply docs/fp_primer.md]: Done

File: docs/fp_primer.md (hunk 9 above)
Comment (zoddicus) [reply docs/fp_primer.md]: Done

File: docs/fp_primer.md (hunk 10 above)
Comment (zoddicus) [reply docs/fp_primer.md]: Done

File: docs/fp_primer.md (hunk 11 above)
Comment (zoddicus) [reply docs/fp_primer.md]: Done

File: docs/fp_primer.md (hunk 13 above)
Comment (zoddicus) [reply docs/fp_primer.md]: Done

File: docs/fp_primer.md (hunk 14 above)
Comment (zoddicus) [reply docs/fp_primer.md]: Done

File: docs/fp_primer.md (hunk 15 above)
Comment (zoddicus) [reply docs/fp_primer.md]: Done

File: docs/fp_primer.md (hunk 16 above)
Comment (zoddicus) [reply docs/fp_primer.md:454]: :+1:

File: docs/fp_primer.md (hunk 17 above)
Comment (zoddicus) [reply docs/fp_primer.md]: Done

File: docs/fp_primer.md (hunk 18 above)
Comment (zoddicus) [reply docs/fp_primer.md]: Done

File: docs/fp_primer.md (hunk 12 above)
Comment (zoddicus) [reply docs/fp_primer.md]: Done

**Comment by ben-clayton** [comment]
This is fantastic. Great job, Ryan!

**Comment by zoddicus** [comment]
@dneto0 are there any further changes that you would like?

File: docs/fp_primer.md [hunk 19]
```diff
@@ -253,8 +260,9 @@ There are 5 different ways that accuracy requirements are defined in the spec:
    The solution to the issue of not scaling with precision of floating point is
    to use units of least precision.
 
-   There are many more formal definitions of ULP, but at a high level it is the
-   difference from a specific value to the next nearest f32 value.
+   ULP(X) is min (b-a) over all pairs (a,b) of representable floating point 
+   numbers such that (a <= X <= b). For a more formal discussion of ULP see
```
Comment (dneto0) [inline docs/fp_primer.md:264]: Sorry, a nit here:  I goofed and we also must have a not equal to b.
It's written correctly in the WGSL spec.

**Review by dneto0 — CHANGES_REQUESTED** [review]
LGTM with one needed fix: in the ULP definition a must be different from b.

**Review by dneto0 — APPROVED** [review]
Meant to give the LGTM.  Expecting the fix tho

File: docs/fp_primer.md (hunk 19 above)
Comment (zoddicus) [reply docs/fp_primer.md]: Done

**Pull request merged**